homeassistant/components/nuki/lock.py

```diff
@@ -147,3 +148,73 @@ def lock_n_go(self, unlatch=False, **kwargs):
         amount of time depending on the lock settings) and relock.
         """
         self._nuki_lock.lock_n_go(unlatch, kwargs)
+
+
+class NukiOpenerEntity(LockEntity):
```

Comment:
Let's create a base class and share much as possible between Lock and Opener. There are too much copy past of the code.

Reply:
Makes sense. I just released pynuki 1.3.7 to make this is a bit more easy.
There's a new NukiBridge.devices property that will return all locks + openers.

Reply:
Thanks, guys. I'm on it.

Reply:
Me too ;)

https://gist.github.com/pschmitt/5ac5d42c68a8f249aaeb18f2b043dfca

Reply:
Awesome, I'll test it. Thanks!

Reply:
I think it's fine to have a base class with and 2 class for the action instead to call isInstance which is slower.

Reply:
Valid point.

Reply:
Something like this then?
[Updated gist](https://gist.github.com/pschmitt/5ac5d42c68a8f249aaeb18f2b043dfca)

Reply:
@pschmitt I know you don't consider it finalized, but I tested your gist and received this:

> may 19 09:30:19 My-Server hass[67628]: 2020-05-19 09:30:19 ERROR (MainThread) [homeassistant.bootstrap] Error setting up integration lock - received exception
may 19 09:30:19 My-Server hass[67628]: Traceback (most recent call last):
may 19 09:30:19 My-Server hass[67628]:   File "/srv/homeassistant/lib/python3.8/site-packages/homeassistant/setup.py", line 50, in async_setup_component
may 19 09:30:19 My-Server hass[67628]:     return await task  # type: ignore
may 19 09:30:19 My-Server hass[67628]:   File "/srv/homeassistant/lib/python3.8/site-packages/homeassistant/setup.py", line 145, in _async_setup_component
may 19 09:30:19 My-Server hass[67628]:     processed_config = await conf_util.async_process_component_config(
may 19 09:30:19 My-Server hass[67628]:   File "/srv/homeassistant/lib/python3.8/site-packages/homeassistant/config.py", line 777, in async_process_component_config
may 19 09:30:19 My-Server hass[67628]:     platform = p_integration.get_platform(domain)
may 19 09:30:19 My-Server hass[67628]:   File "/srv/homeassistant/lib/python3.8/site-packages/homeassistant/loader.py", line 272, in get_platform
may 19 09:30:19 My-Server hass[67628]:     cache[full_name] = importlib.import_module(
may 19 09:30:19 My-Server hass[67628]:   File "/usr/lib/python3.8/importlib/__init__.py", line 127, in import_module
may 19 09:30:19 My-Server hass[67628]:     return _bootstrap._gcd_import(name[level:], package, level)
may 19 09:30:19 My-Server hass[67628]:   File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
may 19 09:30:19 My-Server hass[67628]:   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
may 19 09:30:19 My-Server hass[67628]:   File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
may 19 09:30:19 My-Server hass[67628]:   File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
may 19 09:30:19 My-Server hass[67628]:   File "<frozen importlib._bootstrap_external>", line 783, in exec_module
may 19 09:30:19 My-Server hass[67628]:   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
may 19 09:30:19 My-Server hass[67628]:   File "/home/homeassistant/.homeassistant/custom_components/nuki/lock.py", line 89, in <module>
may 19 09:30:19 My-Server hass[67628]:     class NukiDeviceEntity(LockEntity):
may 19 09:30:19 My-Server hass[67628]:   File "/home/homeassistant/.homeassistant/custom_components/nuki/lock.py", line 109, in NukiDeviceEntity
may 19 09:30:19 My-Server hass[67628]:     def is_locked(self):
may 19 09:30:19 My-Server hass[67628]:   File "/usr/lib/python3.8/abc.py", line 24, in abstractmethod
may 19 09:30:19 My-Server hass[67628]:     funcobj.__isabstractmethod__ = True
may 19 09:30:19 My-Server hass[67628]: AttributeError: attribute '__isabstractmethod__' of 'property' objects is not writable

Reply:
That was just spaghetti code. I haven't tested it myself tbh. Just wanted to drop a few hints.

Reply:
Okay, thanks for the push :)
I did some reading on ABCs, and I've got it working. 
https://gist.github.com/thomkaufmann/bc79f82d4c6e500e2c13c5cf255723e1
Although the actions are immediate, there is a 20 to 30 second delay in the lock receiving the state change from unlocked to locked after locking it, so I'm looking into that. The opener works perfectly.

Reply:
Well that's normal. It is due to the fact that the scan_interval is set [to 15s by default](https://github.com/home-assistant/core/blob/2b38df27668b721ca8d7ca5bb9bd7992f8f9fcfc/homeassistant/helpers/entity_component.py#L29). We are doing polling here after all. The only wait to decrease the delay without implementing webhooks (as described [here](https://github.com/home-assistant/core/pull/35702#issuecomment-629769273)) is to reduce in your config.

Reply:
Okay, that's sensible. I just committed the changes: https://github.com/thomkaufmann/core/commit/e4441c7d9cc008edd0b6efe7df558cdec26d1ba1

Reply:
Nice. Just may want to run black on your files though.

Reply:
Done. Thanks ;)

Reply:
But wait there's more. Please check the CI logs (isort)

Reply:
@thomkaufmann As stated by the CI log you can remove the pass statements on lines 100, 139, 144 and 149. I didn't know that this is allowed either.

Ref: https://github.com/PyCQA/pylint/issues/2208

Reply:
Thanks, @pschmitt!